### PR TITLE
Add passport

### PR DESCRIPTION
**Pull request by youssef910**
Add auth and create user to backend

### CODE REVIEW

File: server/api/general.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import db from "../db";
+
+const passport = require("passport");
+
+
+
+router.get("/health", function (_, response) {
+	db.query("select version()")
+		.then((result) => {
+			if (result.rows.length > 0) {
```
Comment (khaledkz) [inline server/api/general.js:12]: if rows is null this can cause an issue not sure if this scenario can happen, if yes you need to check if rows is not null or undefined

File: server/api/index.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+
+
+import users from "./users";
+import general from "./general";
+
+
+
+
+
```
Comment (khaledkz) [inline server/api/index.js:12]: empty lines

File: server/auth/passport.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+config.jwtFromRequest = ExtractJwt.fromAuthHeaderAsBearerToken("authorization");
+
+passport.use(
+	new JwtStrategy(config, async (jwtPayload, done) => {
+		try {
+			const user = await db.getUserById(jwtPayload.userId);
+			if (user) {
+				return done(null, user);
+			} else {
+				return done(null, false);
```
Comment (ewaDev) [inline server/auth/passport.js:17]: this can be converted into a return done(null, user ? user : false )

File: server/api/general.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
+const express = require("express");
+const router = express.Router();
+import db from "../db";
+
+const passport = require("passport");
+
```
Comment (ewaDev) [inline server/api/general.js:6]: empty lines

File: server/api/users.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+const express = require("express");
+const router = express.Router();
+import { getAllUsers, getUserById }  from "../services/database/users.js";
+
+
+
+
```
Comment (ewaDev) [inline server/api/users.js:7]: lines

File: server/api/users.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+router.get("/", (_, res) => {
+	res.json({ message: "Hello, world!" });
+});
```
Comment (ewaDev) [inline server/api/users.js:10]: is this relevant to code?

File: server/api/users.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+router.get("/get-all", function (_, response) {
```
Comment (ewaDev) [inline server/api/users.js:12]: if you have a users end point to get all users you would just use 

GET apiLink/users. 

It is not restfull to have the action name in the link

File: server/api/users.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+			} else {
+				response.sendStatus(400);
+			}
+		})
+		.catch((ex) => {
+			console.error(ex.message);
+		});
+});
+
+router.get("/:id",(request,response)=>{
```
Comment (ewaDev) [inline server/api/users.js:25]: similarly here you would use

GET apiLink/users/{UserID}

File: server/api/users.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+		})
+		.catch((ex) => {
+			console.error(ex.message);
+		});
+});
+router.get("/:id",(request,response)=>{
+	const id = request.params.id;
+	getUserById(id)
+		.then((result) => {
+			if (result.length > 0) {
```
Comment (ewaDev) [inline server/api/users.js:43]: duplicate oif the above?

File: server/app.js [hunk 10]
```diff
@@ -11,10 +19,11 @@ import {
 } from "./middleware";
 
 const apiRoot = "/api";
-const staticDir = path.join(__dirname, "static");
+// const staticDir = path.join(__dirname, "static");
```
Comment (ewaDev) [inline server/app.js:22]: do we need this comment here?

File: server/services/database/users.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+			}
+		);
+	});
+};
+
+ export const createUser = ({ name, email, pwd }) => {
+    console.log(name,email,pwd)
+	return new Promise((resolve, reject) => {
+		db.query(
+			"INSERT INTO users (name,email, pwd) values ($1, $2,$3) RETURNING id",
```
Comment (ewaDev) [inline server/services/database/users.js:31]: spacing on the query

File: server/services/database/users.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+			[email],
+			(error, result) => {
+				resolve(result.rows[0]);
+			}
+		);
+	});
+};
+
+ export const createUser = ({ name, email, pwd }) => {
+    console.log(name,email,pwd)
```
Comment (ewaDev) [inline server/services/database/users.js:28]: remove console.logs

File: server/api/general.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+				response.sendStatus(400);
+			}
+		})
+		.catch((ex) => {
+			console.error(ex.message);
+		});
+	});
+
+
+/** route is /status/ (status is defined as prefix in index.js) */
```
Comment (ewaDev) [inline server/api/general.js:24]: there is no status in the index.js?

File: server/api/general.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+router.get("/", (req, res) => {
```
Comment (ewaDev) [inline server/api/general.js:25]: This has no use

File: server/api/general.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+router.get(
+	"/protected",
+	passport.authenticate("jwt", { session: false }),
+	(req, res) => {
+		res.send("All good");
+	}
+);
+
+
+
```
Comment (ewaDev) [inline server/api/general.js:42]: lines

File: server/api/users.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+				response.sendStatus(400);
+			}
+		})
+		.catch((ex) => {
+			console.error(ex.message);
+		});
+});
+router.get("/:id",(request,response)=>{
+	const id = request.params.id;
+	getUserById(id)
```
Comment (ewaDev) [inline server/api/users.js:41]: semicolon pls

File: server/app.js [hunk 17]
```diff
@@ -25,9 +34,10 @@ if (app.get("env") === "production") {
 }
 
 app.use(apiRoot, router);
+app.use("/auth",auth);
 
-app.use(express.static(staticDir));
-app.use(pushStateRouting(apiRoot, staticDir));
+// app.use(express.static(staticDir));
+// app.use(pushStateRouting(apiRoot, staticDir));
```
Comment (ewaDev) [inline server/app.js:40]: do we need the comments here?

File: server/auth/routes.js [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+const router = express.Router();
+const jwt = require("jsonwebtoken");
+const db = require("../services/database/users");
+
+/**
+ * Users Login
+ */
+router.post("/login", async (req, res, next) => {
+	const { email, pwd } = req.body;
+	console.log("aksa;ls",email,pwd);
```
Comment (ewaDev) [inline server/auth/routes.js:11]: are these console logs necessary

File: server/auth/routes.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+	console.log(`Login attempt ${email}`);
+	try {
+		const user = await db.getUserByEmail(email);
+
+		if (!email || !pwd || !user || pwd !== user.pwd) {
```
Comment (ewaDev) [inline server/auth/routes.js:16]: I would put this in a variable so it is easier to read

File: server/auth/routes.js [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+			return res.sendStatus(403);
+		}
+
+		const secret = process.env.JWT_SECRET || "your_jwt_secret";
+		const token = jwt.sign({ userId: user.id }, secret);
+		delete user.pwd;
+		delete user.salt;
+		return res.send({ token, user });
+	} catch (e) {
```
Comment (ewaDev) [inline server/auth/routes.js:25]: sometimes you use (e) sometimes you use (ex) - you should decide on one and stick to it

File: server/auth/routes.js [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+		console.error(e);
+		next(e);
+	}
+});
+
+/**
+ * Users Registration
+ */
+router.post("/register", async (req, res, next) => {
```
Comment (ewaDev) [inline server/auth/routes.js:34]: maybe im not understanding this, but this should sit in the USER controller. so that the path is 

POST ApiLink/Users

File: server/services/database/users.js [hunk 22]
```diff
@@ -0,0 +1,56 @@
+import db from "../../db";
```
Comment (ewaDev) [inline server/services/database/users.js:1]: Its fine for a small project, but usually you should have any code which interacts with the DB, put in the "repository" layer 

so Controller -> Service -> Repository.

File: server/auth/routes.js [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+		delete user.salt;
+		return res.send({ token, user });
+	} catch (e) {
+		console.error(e);
+		next(e);
+	}
+});
+
+/**
+ * Users Registration
```
Comment (ewaDev) [inline server/auth/routes.js:32]: no need for comment

**Review by ewaDev — CHANGES_REQUESTED** [review]
some of the paths need work but great work otherwise! 

Also not sure if this is in scope, but there are no integration or unit tests for these... We should really be writing tests
